src/main/java/hudson/remoting/Channel.java

```diff
@@ -1006,9 +1014,12 @@ public void checkRoles(RoleChecker checker) throws SecurityException {
      * @since 1.299
      */
     public synchronized void join(long timeout) throws InterruptedException {
-        long start = System.currentTimeMillis();
-        while(System.currentTimeMillis()-start<timeout && (inClosed==null || outClosed==null))
-            wait(timeout+start-System.currentTimeMillis());
+        long now = System.nanoTime();
+        long end = now + (timeout * 1000L);
+        while ((end - now > 0L) && (inClosed == null || outClosed == null)) {
+            wait((end - now) / 1000L);
```

Comment:
units again converting nanoseconds to milliseconds... also since you have an actual Lock it would make more sense to add a Condition and use Condition.awaitNanos() to reduce the running total rather than relying on a wall clock no matter how monotonically increasing it claims to be